File: Yr2022/day_twenty.py
from time import perf_counter_ns
from tqdm import tqdm


def shift_em(orig_list, num_shifts=10, decrypt_key = 811589153):
    pair_list = list(enumerate(val * decrypt_key for val in orig_list))
    
    for pair in pair_list * num_shifts:
        start_idx = pair_list.index(pair)
        pair_list.pop(start_idx)
        new_idx = (start_idx+pair[1]) % len(pair_list)
        pair_list.insert( new_idx , pair)
       
    zero_pos = [end_idx for end_idx, (_, val) in enumerate(pair_list) if val == 0][0]
       
    sum_list = []
    for point in [1000, 2000, 3000]:
        new_point_index = (zero_pos + point) % len(pair_list)
        sum_list.append(pair_list[new_point_index][1])
    print("SUM LIST ", zero_pos,  sum_list)
    return sum(sum_list)
    


def apply_shifts(filename: str):
    cmd_lines= []
    with open(filename, 'r') as file:
        cmd_lines = file.readlines()
    
    orig_list = [int(x.strip()) for x in cmd_lines]
        
    part_one = shift_em(orig_list, num_shifts=1, decrypt_key=1)    
    part_two = shift_em(orig_list, num_shifts=10, decrypt_key=811589153)
    
    return part_one, part_two
  

if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_twenty.txt"     
    #file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\example.txt"     
    
    part_one, part_two = apply_shifts(file_loc)
    
    print(f"(Part One): {part_one}     (Part Two): {part_two}")   
    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")  

File: Yr2022/src/day_twenty.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;


fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_twenty.txt";
    //let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\example.txt";
    let (part_one, part_two) = apply_shifts(file_name);
    println!("Part One: {}  Part Two:  {}", part_one, part_two);
    let duration = start.elapsed();    
    println!("Time taken to execute: {} microseconds", duration.as_micros());    
}



fn shift_em(orig_vec: &Vec<i64>, num_shifts: usize, decrypt_key: i64) -> i64 {
    let mut pair_vec: Vec< (usize, i64) > = Vec::new();

    for (idx, val) in orig_vec.iter().enumerate() {
        pair_vec.push( (idx, val*decrypt_key));
    }   
    for _rep in 0..num_shifts {
        for pair_idx in 0..pair_vec.len() {
            let start_idx = pair_vec.iter().position(|&p| p.0 == pair_idx).unwrap();
            let pair = pair_vec[start_idx].clone();
            pair_vec.remove(start_idx);
            let new_idx = (start_idx as i64 + pair.1).rem_euclid(pair_vec.len() as i64) as usize;
            pair_vec.insert(new_idx, pair);
        }
    }

    let zero_pos = pair_vec.iter().position(|p| p.1 == 0).unwrap();    
    let check_pos: Vec<usize> = vec![1000, 2000, 3000];
    return check_pos.iter().map(|pos| {
        let new_pos_index = (zero_pos + pos).rem_euclid(pair_vec.len());
        pair_vec[new_pos_index].1
    } ).sum();
}


fn apply_shifts(my_file_name: &str) -> (i64, i64){        
    let file = File::open(my_file_name).expect("File does not exist");
    let buf_lines = BufReader::new(file).lines();
    let line_vec: Vec<String> = buf_lines.map(|l| l.expect("Fail to parse")).collect();
    let orig_vec: Vec<i64> = line_vec.iter().map(|l| l.trim().parse::<i64>().unwrap() ).collect();

    let part_one = shift_em(&orig_vec, 1, 1);
    let part_two = shift_em(&orig_vec, 10, 811589153);
    
    return (part_one, part_two);
}


